fedbiomed/transport/client.py

```diff
@@ -485,113 +575,120 @@ def __init__(
+                await post_noretry_function(*args)
+
+
+    def _message_deadline_exceeded(self):
+        """Sender issues warning when researcher does not complete request before deadline
         """
+        logger.warning(
+            "Researcher not answering after timeout, looks like server failure or disconnect. "
+            "Discard message.")
 
-        # While loop retires to send if first one fails to send the result
-        while True:
-            try:
-                await self._get(callback)
-            except grpc.aio.AioRpcError as exp:
-                match exp.code():
-                    case grpc.StatusCode.DEADLINE_EXCEEDED:
-                        await self._on_status_change(ClientStatus.DISCONNECTED)
-                        logger.warning(
-                            "Researcher not answering after timeout, looks like server failure or disconnect. "
-                            "Discard message.")
-                        self._queue.task_done()
-                        self._retry_count = 0
-                    case grpc.StatusCode.UNAVAILABLE:
-                        await self._on_status_change(ClientStatus.DISCONNECTED)
-                        logger.info(
-                            "Researcher server is not available, will retry connect in "
-                            f"{GRPC_CLIENT_CONN_RETRY_TIMEOUT} seconds")
-                        await asyncio.sleep(GRPC_CLIENT_CONN_RETRY_TIMEOUT)
-                        await self._channels.connect(self._stub_type)
-                        self._retry_count += 1
-                    case grpc.StatusCode.UNKNOWN:
-                        await self._on_status_change(ClientStatus.FAILED)
-                        logger.error("Unexpected error raised by researcher gRPC server. This is probably due to "
-                                     f"bug on the researcher side: {exp}")
-                    case _:
-                        await self._on_status_change(ClientStatus.FAILED)
-
-            except Exception as exp:
-                await self._on_status_change(ClientStatus.FAILED)
-                raise FedbiomedCommunicationError(
-                    f"{ErrorNumbers.FB628}: Sender has stopped due to unknown reason: {exp}") from exp
-
-            except GeneratorExit as exp:
-                await self._on_status_change(ClientStatus.FAILED)
-                raise FedbiomedCommunicationError(
-                    f"{ErrorNumbers.FB628}: Sender has stopped due to unexpected gRPC abort: {exp}") from exp
-
-
-    async def _get(self, callback: Optional[Callable] = None) -> None:
+
+    async def _call_researcher(self, callback: Optional[Callable] = None) -> None:
         """Gets task result from the queue.
 
         Args:
             callback: Callback to execute once a task is received
         """
+        if self._retry_count == 0:
+            # only pick a new message if not retrying to send
+            self._retry_msg = await self._queue.get()
```

Comment:
You are right this is a job/item not a message, and this is a bit confusing (felt the same ...). I did not change it because this is also the terminology used in `Sender._call_researcher()` ... but will change it for both to apply your recommandation